cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go

```diff
@@ -21,38 +21,134 @@ import (
 	"regexp"
 	"strconv"
 	"strings"
+	"sync"
+	"time"
 
+	aws_util "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
 	"k8s.io/contrib/cluster-autoscaler/cloudprovider"
 	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	kube_client "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5"
 )
 
 // AwsCloudProvider implements CloudProvider interface.
 type AwsCloudProvider struct {
 	awsManager *AwsManager
+	asgMutex   sync.Mutex
 	asgs       []*Asg
 }
 
 // BuildAwsCloudProvider builds CloudProvider implementation for AWS.
-func BuildAwsCloudProvider(awsManager *AwsManager, specs []string) (*AwsCloudProvider, error) {
+func BuildAwsCloudProvider(awsManager *AwsManager, specs []string, kubeClient kube_client.Interface, autoDiscoverNodeGroup bool, scanInterval time.Duration) (*AwsCloudProvider, error) {
 	aws := &AwsCloudProvider{
 		awsManager: awsManager,
 		asgs:       make([]*Asg, 0),
 	}
-	for _, spec := range specs {
-		if err := aws.addNodeGroup(spec); err != nil {
-			return nil, err
+	if autoDiscoverNodeGroup {
+		go func() {
+			for {
+				select {
+				case <-time.After(scanInterval):
+					{
+						selector := "master!=true"
+						listAll := apiv1.ListOptions{LabelSelector: selector}
+						nl, err := kubeClient.Core().Nodes().List(listAll)
+						if err != nil {
+							glog.Errorf("err: %s\n", err)
+						}
+
+						var nodesList []*string
+
+						for _, n := range nl.Items {
+							glog.V(4).Infof("Considering node: %s", n.Name)
+							nodesList = append(nodesList, &n.Name)
+						}
+						AutoDiscoverNodeGroup(aws, nodesList)
```

Comment:
@Raffo The `Autoscaler` interface is recently introduced to the code-base. Would it be possible to utilize it here so that we can achieve it without goroutines?
https://github.com/kubernetes/contrib/blob/master/cluster-autoscaler/core/autoscaler.go#L38

Just a rough idea but add an `Autoscaler` implementation. The implementation, when `currentTime > lastFetchTime + delay`, fetches autoscaling groups from AWS and update node groups like we do for [DynamicAutoscaler](https://github.com/kubernetes/contrib/blob/master/cluster-autoscaler/core/dynamic_autoscaler.go#L125).